### PR TITLE
use the accurate docContext and persist repo visibility using local storage

### PR DESCRIPTION
**Pull request by hitesh-1997**
The PR improves the following:
1. When calculating the `diff` over time for suggestion, we were using `document.getText`. Analyzing offline telemetry we see in ~20% we get some extra chars if user types between calculating `docContext` and `loaded` call. We leverage already computed `completePrefix` and `completeSuffix` to compute the text, which is accurate.
2. We were using in-memory cache for calculating `repoVisibility` before. But without passing token, we frequently hit rate limit for querying github api and cached tokens are evicted from in-memory cache when user refresh. This PR stores the tokens in `localStorage` instead of in-memory store to cache the repo visibility.
3. Adds the `code qwen` and `deepseek` model optimized for long prompts.

## Test plan
1. Added unit tests.
2. Manual Testing: Adding debugger and refreshing to see if the token value is persisted/ deleted from local storage.

### CODE REVIEW

File: vscode/src/completions/inline-completion-item-provider.ts [hunk 1]
```diff
@@ -843,10 +843,13 @@ export class InlineCompletionItemProvider
                 takeSuggestWidgetSelectionIntoAccount,
                 undefined
             )
+            completion.requestParams.docContext.position
             if (isStillVisible) {
                 suggestionEvent.markAsRead({
                     document: invokedDocument,
                     position: invokedPosition,
+                    docPrefix: completion.requestParams.docContext.completePrefix,
+                    docSuffix: completion.requestParams.docContext.completeSuffix,
```
Comment (valerybugakov) [inline vscode/src/completions/inline-completion-item-provider.ts:852]: Let's avoid renaming things as much as possible to make it easier to trace the data flow. It's unclear if the `markAsRead` implementation expects `completePrefix` or `prefix` based on its function parameters. 

```suggestion
                    completePrefix: completion.requestParams.docContext.completePrefix,
                    completeSuffix: completion.requestParams.docContext.completeSuffix,
```

File: vscode/src/completions/logger.ts [hunk 2]
```diff
@@ -775,7 +775,9 @@ function getInlineContextItemContext(
 function suggestionDocumentDiffTracker(
     interactionId: CompletionAnalyticsID,
     document: vscode.TextDocument,
-    position: vscode.Position
+    position: vscode.Position,
+    docPrefix: string,
+    docSuffix: string
```
Comment (valerybugakov) [inline vscode/src/completions/logger.ts:780]:
```suggestion
    completePrefix: string,
    completeSuffix: string
```

File: vscode/src/completions/model-helpers/codeqwen.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    getOllamaPrompt(promptContext: GetOllamaPromptParams): PromptString {
+        const { context, currentFileNameComment, prefix, suffix } = promptContext
+
+        const infillPrefix = context.concat(currentFileNameComment, prefix)
+
+        return ps`<|fim_prefix|>${infillPrefix}<|fim_suffix|>${suffix}<|fim_middle|>`
+    }
+
+    getOllamaRequestOptions(isMultiline: boolean): OllamaGenerateParameters {
+        const params = {
```
Comment (valerybugakov) [inline vscode/src/completions/model-helpers/codeqwen.ts:35]: Did you have a chance to test it locally with Ollama?

File: vscode/src/completions/model-helpers/codeqwen.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        return ps`<|file_sep|>${PromptString.fromDisplayPath(snippet.uri)}\n${content}`
+    }
+
+    formatPrompt(params: FormatPromptParams): PromptString {
+        // Prompt format for CodeQwen in technical report: https://arxiv.org/pdf/2409.12186
+        const { intro, prefix, suffix, repoName, fileName } = params
+        let introPrefix = ps``
+        if (intro.length > 0) {
+            introPrefix = ps`${intro}\n`
+        }
```
Comment (valerybugakov) [inline vscode/src/completions/model-helpers/codeqwen.ts:75]: Is this already handled in `formatIntroSnippets`?

**Review by valerybugakov — APPROVED** [review]

File: vscode/src/completions/inline-completion-item-provider.ts [hunk 5]
```diff
@@ -843,10 +843,13 @@ export class InlineCompletionItemProvider
                 takeSuggestWidgetSelectionIntoAccount,
                 undefined
             )
+            completion.requestParams.docContext.position
```
[the remark this answers is not in the document]
Comment (hitesh-1997) [reply vscode/src/completions/inline-completion-item-provider.ts:846]: `position` is being used in the `persistenceTracker` to calculate range around trigger point.

File: vscode/src/completions/inline-completion-item-provider.ts (hunk 1 above)
Comment (hitesh-1997) [reply vscode/src/completions/inline-completion-item-provider.ts]: makes sense, changed as per suggestion

File: vscode/src/completions/logger.ts (hunk 2 above)
Comment (hitesh-1997) [reply vscode/src/completions/logger.ts]: done

File: vscode/src/completions/model-helpers/codeqwen.ts (hunk 3 above)
Comment (hitesh-1997) [reply vscode/src/completions/model-helpers/codeqwen.ts]: ah, sorry didn't test with Ollama
for now, removed the ollama options from the file, as we want to A/B test the model itself.

File: vscode/src/completions/model-helpers/codeqwen.ts (hunk 4 above)
Comment (hitesh-1997) [reply vscode/src/completions/model-helpers/codeqwen.ts]: yes, removed from here

File: vscode/src/completions/inline-completion-item-provider.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply vscode/src/completions/inline-completion-item-provider.ts:846]: @hitesh-1997, does this line do anything? It looks like a reference to the existing field on the `docContext` object and this line itself doesn't affect its presence.

[the remark this answers is not in the document]
Comment (hitesh-1997) [reply vscode/src/completions/inline-completion-item-provider.ts:846]: Ah yes, sorry I looked at some other file.
raised this PR to clean this: https://github.com/sourcegraph/cody/pull/5793/files

**Pull request merged**